datafusion/functions-aggregate/benches/array_agg.rs

```diff
@@ -0,0 +1,186 @@
+
+use arrow::array::{Array, ArrayRef, ArrowPrimitiveType, AsArray, ListArray};
+use arrow::datatypes::Int64Type;
+use arrow::util::bench_util::create_primitive_array;
+use arrow_schema::Field;
+use criterion::{black_box, criterion_group, criterion_main, Criterion};
+use datafusion_expr::Accumulator;
+use datafusion_functions_aggregate::array_agg::ArrayAggAccumulator;
+
+use arrow::util::test_util::seedable_rng;
+use arrow_buffer::{NullBufferBuilder, OffsetBuffer};
+use rand::distributions::{Distribution, Standard};
+use rand::Rng;
+
+fn merge_batch_bench(c: &mut Criterion, name: &str, values: ArrayRef) {
+    let list_item_data_type = values.as_list::<i32>().values().data_type().clone();
+    c.bench_function(name, |b| {
+        b.iter(|| {
+            #[allow(clippy::unit_arg)]
+            black_box(
+                ArrayAggAccumulator::try_new(&list_item_data_type)
+                    .unwrap()
+                    .merge_batch(&[values.clone()])
+                    .unwrap(),
+            )
+        })
+    });
+}
+
+/// Create List array with the given item data type, null density, null locations and non zero length lists density
+/// Creates an random (but fixed-seeded) array of a given size and null density
+pub fn create_list_array<T>(
+    size: usize,
+    null_density: f32,
+    non_zero_length_lists_probability: f32,
+) -> ListArray
+where
+    T: ArrowPrimitiveType,
+    Standard: Distribution<T::Native>,
+{
+    let mut nulls_builder = NullBufferBuilder::new(size);
+    let mut rng = seedable_rng();
+
+    let offsets = OffsetBuffer::from_lengths((0..size).map(|_| {
+        let is_null = rng.gen::<f32>() < null_density;
+
+        let mut length = rng.gen_range(1..10);
+
+        if is_null {
+            nulls_builder.append_null();
+
+            if rng.gen::<f32>() > non_zero_length_lists_probability {
+                length = 0;
+            }
+        } else {
+            nulls_builder.append_non_null();
+        }
+
+        length
+    }));
+
+    let length = *offsets.last().unwrap() as usize;
+
+    let values = create_primitive_array::<T>(length, 0.0);
+
+    let field = Field::new_list_field(T::DATA_TYPE, true);
+
+    ListArray::new(
+        Arc::new(field),
+        offsets,
+        Arc::new(values),
+        nulls_builder.finish(),
+    )
+}
+
+fn array_agg_benchmark(c: &mut Criterion) {
+    let values = Arc::new(create_list_array::<Int64Type>(8192, 0.0, 0.0)) as ArrayRef;
+    merge_batch_bench(c, "array_agg i64 merge_batch no nulls", values);
+
+    let values = Arc::new(create_list_array::<Int64Type>(8192, 1.0, 0.0)) as ArrayRef;
+    merge_batch_bench(
+        c,
+        "array_agg i64 merge_batch all nulls, 100% of nulls point to a zero length array",
+        values,
+    );
+
+    let values = Arc::new(create_list_array::<Int64Type>(8192, 1.0, 0.1)) as ArrayRef;
+    merge_batch_bench(
+        c,
+        "array_agg i64 merge_batch all nulls, 90% of nulls point to a zero length array",
```

Comment:
this comment says 90 but the function call is `create_list_array::<Int64Type>(8192, 1.0, 0.1)` (which is 100 right?)

Reply:
Or is the idea that 10% nulls point at a non zero length array 🤔

Reply:
all null because null density (2nd arg) is 1.0, and 90% of nulls point to a zero length array, because the 3rd argument is the probability for non-zero length so it's 10% and 100 - 10 is 90% hence the 90% of nulls

I'll rename the argument in the function to be probability for zero length array

Reply:
Thanks -- that makes a lot more sense to me and I found it easier to understand. 🙏